configurator/src/main/java/org/apache/felix/configurator/impl/model/State.java

```diff
@@ -39,9 +39,8 @@ public class State extends AbstractState implements Serializable {
 
     public static final String FILE_NAME = "state.ser";
 
-    private Map<Long, Long> bundlesLastModified = new HashMap<>();
-
-    private Map<Long, Long> bundlesConfigAdminBundleId = new HashMap<>();
+    private Map<Long, Long> bundlesLastModified = new ConcurrentHashMap<>();
```

Comment:
I don't get why the default implementation of serializable is not used here to read/write the maps so you can make these final fields.

Reply:
Its used for versioning of the serialized format, it makes it easier to make any change to this class in future versions. It might be a little bit overkill here

Reply:
@cziegeler Could you kindly elaborate what I am missing from the PR?

Reply:
The Map type changing is what triggered my comment about serialization.  The concern is that a previous data stream would be storing a HashMap instance and when you do the read on restoring it will construct a HashMap instance back.  So in that case you need to read into a Map local instance for the call to `Map<Long, Long>) in.readObject();` then check the instance of the Map.  If it is not a ConcurrentHashMap then you need to copy the content of the read map into a new ConcurrentHashMap instance.

But it seems there is other issues with thread safety for the `private volatile Set<String> initialHashes` set.  Is that always set to a ConcurrentHashSet or something similar to ensure it is thread safe?

Reply:
Tom is right, you need to create the right type on deseralization.
But, it might also be that the synchronization is on the wrong level and instead of making the properties of this class thread safe take care of syncing on a higher level. I think, that was the intention, but I can't really remember the details

Reply:
@tjwatson @cziegeler Thanks a lot for the clarifications 👍  It's my bad since I didn't notice it earlier. I will add the necessary changes and update the PR accordingly.

Reply:
@tjwatson @cziegeler I have updated the PR. Could you kindly have another look?